storage/storage.go

```diff
@@ -13,3 +20,120 @@ func NewStorage(bucket pail.Bucket) Storage {
+	}
+	return
+}
+
+func (storage *Storage) getAllChunks(context context.Context, buildId string) ([]LogChunkInfo, error) {
+	buildPrefix := fmt.Sprintf("/%s/", buildId)
+
+	iterator, listErr := storage.bucket.List(context, buildPrefix)
+	buildChunks := []LogChunkInfo{}
+	if listErr != nil {
+		return nil, listErr
+	}
+	for iterator.Next(context) {
+		if strings.HasSuffix(iterator.Item().Name(), "metadata.json") {
+			continue
+		}
+		if strings.Contains(iterator.Item().Name(), "/tests/") {
+			keyParts := strings.Split(iterator.Item().Name(), "/")
+			buildId := keyParts[1]
+			test_id := keyParts[3]
+			name := keyParts[4]
+			start, end, numLines, nameErr := parseName(name)
+			if nameErr != nil {
+				return nil, nameErr
+			}
+			buildChunks = append(buildChunks, LogChunkInfo{
+				BuildId:  buildId,
+				TestId:   test_id,
+				Start:    start,
+				End:      end,
+				NumLines: int(numLines),
+			})
+		} else {
+			keyParts := strings.Split(iterator.Item().Name(), "/")
+			buildId := keyParts[1]
+			name := keyParts[2]
+			start, end, numLines, nameErr := parseName(name)
+			if nameErr != nil {
+				return nil, nameErr
+			}
+			buildChunks = append(buildChunks, LogChunkInfo{
+				BuildId:  buildId,
+				TestId:   "",
+				Start:    start,
+				End:      end,
+				NumLines: int(numLines),
+			})
+		}
+	}
+	return buildChunks, nil
+}
+
+func (storage *Storage) GetTestLogLines(context context.Context, buildId string, testId string) (LogIterator, error) {
+	chunks, err := storage.getAllChunks(context, buildId)
+	if err != nil {
+		return nil, err
+	}
+
+	testChunks := []LogChunkInfo{}
+	for i := 0; i < len(chunks); i++ {
+		// Find our test id
+		if chunks[i].TestId == testId {
+			testChunks = append(testChunks, chunks[i])
+		}
+	}
+	sort.Slice(testChunks, func(i, j int) bool {
+		return testChunks[i].Start.Before(testChunks[j].Start)
+	})
+
+	testTimeRange := TimeRange{
+		StartAt: testChunks[0].Start,
+		EndAt:   testChunks[len(testChunks)-1].End,
+	}
+
+	testChunkIterator := NewBatchedLogIterator(storage.bucket, testChunks, 4, testTimeRange)
+
+	buildChunks := []LogChunkInfo{}
+	for i := 0; i < len(chunks); i++ {
+		// Include any build logs that are in the time range of our test
+		chunkTimeRange := TimeRange{
+			StartAt: chunks[i].Start,
+			EndAt:   chunks[i].End,
+		}
+		// check if the global build chunk's time range intersects the test's time range, and if so
+		// add it to our list of build chunks, but constrained to the test's time range to only
+		// include entries during that time.
+		if chunks[i].TestId == "" && testTimeRange.Intersects(chunkTimeRange) {
+			buildChunks = append(buildChunks, chunks[i])
+		}
+	}
+
+	sort.Slice(buildChunks, func(i, j int) bool {
+		return buildChunks[i].Start.Before(buildChunks[j].Start)
+	})
+	buildChunkIterator := NewBatchedLogIterator(storage.bucket, buildChunks, 4, testTimeRange)
+
+	// Merge everything together
+	return NewMergingIterator(testChunkIterator, buildChunkIterator), nil
```

Comment:
if we just have two streams do we need to use the heap implementation? (it would seem merging two streams has the advantage of being O(n+m) vs. the heap is O(n log n + m log m))

Reply:
We don't need to use the heap implementation, but it was already written to use the Iterator interface and has been tested in production code, which is nice.

As far as runtime complexity, that "log(N)" component is dependent on heap operations being logN. In the case of the MergingIterator, that is going to be proportional to the size of the heap, which is just the constant (2) since it is a heap of the two iterators supplied. That brings the algorithmic complexity down do O(n + m). Basically since we're not actually putting all the log chunks or log lines on a heap, the heap operation complexity should be constant, and not scale up, and we do one heap operation per log line, so constant O(c * (n + m) ) = O(n+m).

That said, it's a fair bit of additional fixed pointless computational overhead per log line, but I'm guessing that that is dwarfed by the overhead of network handling, processing strings, etc.

Reply:
oh wow! That makes sense.